File: examples/external-packages/au_eftex/au_eftex/iso8583.py
"""Bounded AU EFTEX ISO 8583:1993 parser and encoder.

Only the fields declared in :data:`FIELD_SCHEMA` are understood. Binary
payment fields remain opaque blobs; this module does not interpret PIN, MAC,
or private field contents.
"""

from __future__ import annotations

import base64
import binascii
from dataclasses import dataclass
from typing import Literal, TypedDict, Union


class StringValue(TypedDict):
    type: Literal["string"]
    value: str


class IntValue(TypedDict):
    type: Literal["int"]
    value: str


class BlobValue(TypedDict):
    type: Literal["blob"]
    value_base64: str


DocumentValue = Union[StringValue, IntValue, BlobValue]
Document = dict[str, DocumentValue]

FieldKind = Literal[
    "fixed_ascii",
    "fixed_amount",
    "fixed_blob",
    "ll_ascii",
    "ll_blob",
    "lll_ascii",
    "lll_blob",
    "llll_ascii",
    "llll_blob",
]
DocumentType = Literal["string", "int", "blob"]


@dataclass(frozen=True, slots=True)
class FieldSpec:
    number: int
    name: str
    kind: FieldKind
    length: int
    document_type: DocumentType


FIELD_SCHEMA: tuple[FieldSpec, ...] = (
    FieldSpec(0, "message_type", "fixed_ascii", 4, "string"),
    FieldSpec(2, "primary_account_number", "ll_ascii", 19, "string"),
    FieldSpec(3, "processing_code", "fixed_ascii", 6, "string"),
    FieldSpec(4, "amount", "fixed_amount", 12, "int"),
    FieldSpec(7, "transmission_time", "fixed_ascii", 10, "string"),
    FieldSpec(11, "stan", "fixed_ascii", 6, "string"),
    FieldSpec(12, "local_transaction_time", "fixed_ascii", 12, "string"),
    FieldSpec(14, "expiration_date", "fixed_ascii", 4, "string"),
    FieldSpec(22, "pos_data_code", "fixed_ascii", 15, "string"),
    FieldSpec(23, "card_sequence_number", "fixed_ascii", 3, "string"),
    FieldSpec(24, "function_code", "fixed_ascii", 3, "string"),
    FieldSpec(25, "message_reason_code", "fixed_ascii", 4, "string"),
    FieldSpec(28, "reconciliation_date", "fixed_ascii", 6, "string"),
    FieldSpec(29, "reconciliation_indicator", "fixed_ascii", 3, "string"),
    FieldSpec(35, "track_2_data", "ll_ascii", 37, "string"),
    FieldSpec(37, "retrieval_reference_number", "fixed_ascii", 12, "string"),
    FieldSpec(38, "approval_code", "fixed_ascii", 6, "string"),
    FieldSpec(39, "action_code", "fixed_ascii", 3, "string"),
    FieldSpec(40, "service_code", "fixed_ascii", 3, "string"),
    FieldSpec(41, "terminal_id", "fixed_ascii", 8, "string"),
    FieldSpec(42, "card_acceptor_id", "fixed_ascii", 15, "string"),
    FieldSpec(46, "amounts_fees", "lll_ascii", 204, "string"),
    FieldSpec(48, "additional_private", "llll_blob", 9_999, "blob"),
    FieldSpec(49, "currency", "fixed_ascii", 3, "string"),
    FieldSpec(50, "reconciliation_currency", "fixed_ascii", 3, "string"),
    FieldSpec(52, "pin_data", "fixed_blob", 8, "blob"),
    FieldSpec(53, "security_control_information", "ll_ascii", 48, "string"),
    FieldSpec(54, "additional_amounts", "lll_ascii", 120, "string"),
    FieldSpec(55, "icc_data", "lll_blob", 512, "blob"),
    FieldSpec(56, "original_data_elements", "ll_ascii", 31, "string"),
    FieldSpec(63, "reserved_private", "lll_ascii", 800, "string"),
    FieldSpec(64, "message_authentication_code", "fixed_blob", 8, "blob"),
    FieldSpec(74, "credits_number", "fixed_ascii", 10, "string"),
    FieldSpec(75, "credits_reversal_number", "fixed_ascii", 10, "string"),
    FieldSpec(76, "debits_number", "fixed_ascii", 10, "string"),
    FieldSpec(77, "debits_reversal_number", "fixed_ascii", 10, "string"),
    FieldSpec(81, "authorisations_number", "fixed_ascii", 10, "string"),
    FieldSpec(86, "credits_amount", "fixed_ascii", 16, "string"),
    FieldSpec(87, "credits_reversal_amount", "fixed_ascii", 16, "string"),
    FieldSpec(88, "debits_amount", "fixed_ascii", 16, "string"),
    FieldSpec(89, "debits_reversal_amount", "fixed_ascii", 16, "string"),
    FieldSpec(90, "authorisations_reversal_number", "fixed_ascii", 10, "string"),
    FieldSpec(97, "net_reconciliation_amount", "fixed_ascii", 17, "string"),
    FieldSpec(109, "credits_fee_amounts", "ll_ascii", 84, "string"),
    FieldSpec(110, "debits_fee_amounts", "ll_ascii", 84, "string"),
    FieldSpec(123, "receipt_data", "lll_ascii", 999, "string"),
    FieldSpec(124, "display_data", "lll_ascii", 999, "string"),
    FieldSpec(128, "message_authentication_code_extended", "fixed_blob", 8, "blob"),
)

_MESSAGE_TYPE = FIELD_SCHEMA[0]
_DATA_FIELDS = FIELD_SCHEMA[1:]
_FIELDS_BY_NUMBER = {field.number: field for field in _DATA_FIELDS}
_FIELDS_BY_NAME = {field.name: field for field in FIELD_SCHEMA}


def message_length(message: bytes) -> int | None:
    """Return one complete message length, or ``None`` when more bytes are needed."""

    _require_bytes(message)
    if len(message) < _MESSAGE_TYPE.length:
        return None
    _read_mti(message[:4])
    if len(message) < 12:
        return None

    primary_bitmap = message[4:12]
    bitmap_bytes = 16 if _bitmap_has(primary_bitmap, 1) else 8
    if len(message) < 4 + bitmap_bytes:
        return None
    bitmap = message[4 : 4 + bitmap_bytes]
    _validate_bitmap(bitmap)
    offset = 4 + bitmap_bytes
    maximum = 128 if bitmap_bytes == 16 else 64
    for number in range(2, maximum + 1):
        if number == 65:
            continue
        if not _bitmap_has(bitmap, number):
            continue
        field = _FIELDS_BY_NUMBER[number]
        payload_length, prefix_length = _field_length(message, offset, field)
        if payload_length is None:
            return None
        end = offset + prefix_length + payload_length
        if len(message) < end:
            return None
        if field.document_type != "blob":
            value = _decode_ascii(
                message[offset + prefix_length : end], field.name
            )
            if field.kind == "fixed_amount":
                _require_digits(value, field.name)
        offset = end
    return offset


def decode_message(message: bytes) -> Document:
    """Decode one complete profile message into tagged document values."""

    expected_length = message_length(message)
    if expected_length is None:
        raise ValueError("ISO 8583 message is incomplete")
    if expected_length != len(message):
        raise ValueError("ISO 8583 message has trailing bytes not described by bitmap")

    document: Document = {
        "message_type": {"type": "string", "value": _read_mti(message[:4])}
    }
    primary_bitmap = message[4:12]
    bitmap_bytes = 16 if _bitmap_has(primary_bitmap, 1) else 8
    bitmap = message[4 : 4 + bitmap_bytes]
    offset = 4 + bitmap_bytes
    maximum = 128 if bitmap_bytes == 16 else 64
    for number in range(2, maximum + 1):
        if number == 65:
            continue
        if not _bitmap_has(bitmap, number):
            continue
        field = _FIELDS_BY_NUMBER[number]
        payload_length, prefix_length = _field_length(message, offset, field)
        assert payload_length is not None
        start = offset + prefix_length
        end = start + payload_length
        payload = message[start:end]
        if field.document_type == "blob":
            document[field.name] = {
                "type": "blob",
                "value_base64": base64.b64encode(payload).decode("ascii"),
            }
        else:
            value = _decode_ascii(payload, field.name)
            document[field.name] = (
                {"type": "int", "value": _canonical_integer(value, field.name)}
                if field.document_type == "int"
                else {"type": "string", "value": value}
            )
        offset = end
    return document


def encode_message(document: Document) -> bytes:
    """Encode a tagged document as one profile message without a wire header."""

    if not isinstance(document, dict):
        raise TypeError("document must be a dict")
    unknown = sorted(set(document) - set(_FIELDS_BY_NAME))
    if unknown:
        raise ValueError(f"unknown document field: {unknown[0]}")
    if "message_type" not in document:
        raise ValueError("message_type is required")

    mti = _require_string_value(document["message_type"], "message_type")
    if len(mti) != 4:
        raise ValueError("message_type must contain exactly 4 ASCII digits")
    _require_digits(mti, "message_type")
    mti_bytes = _encode_ascii(mti, "message_type")

    present = [field for field in _DATA_FIELDS if field.name in document]
    has_secondary = any(field.number > 64 for field in present)
    bitmap = bytearray(16 if has_secondary else 8)
    if has_secondary:
        _bitmap_set(bitmap, 1)
    encoded_fields: list[bytes] = []
    for field in present:
        _bitmap_set(bitmap, field.number)
        encoded_fields.append(_encode_field(document[field.name], field))
    return b"".join([mti_bytes, bytes(bitmap), *encoded_fields])


def _validate_bitmap(bitmap: bytes) -> None:
    if len(bitmap) not in {8, 16}:
        raise ValueError("ISO 8583 bitmap must contain 8 or 16 bytes")
    if len(bitmap) == 16 and _bitmap_has(bitmap, 65):
        raise ValueError("tertiary bitmap is not supported by the AU EFTEX profile")
    maximum = 128 if len(bitmap) == 16 else 64
    for number in range(2, maximum + 1):
        if number == 65:
            continue
        if _bitmap_has(bitmap, number) and number not in _FIELDS_BY_NUMBER:
            raise ValueError(f"DE{number} is not supported by the AU EFTEX profile")


def _field_length(
    message: bytes, offset: int, field: FieldSpec
) -> tuple[int | None, int]:
    prefix_length = {
        "ll_ascii": 2,
        "ll_blob": 2,
        "lll_ascii": 3,
        "lll_blob": 3,
        "llll_ascii": 4,
        "llll_blob": 4,
    }.get(field.kind, 0)
    if prefix_length == 0:
        return field.length, 0
    if len(message) < offset + prefix_length:
        return None, prefix_length
    prefix = _decode_ascii(
        message[offset : offset + prefix_length], f"{field.name} length"
    )
    _require_digits(prefix, f"{field.name} length")
    length = int(prefix)
    if length > field.length:
        raise ValueError(f"{field.name} length exceeds profile maximum")
    return length, prefix_length


def _encode_field(value: DocumentValue, field: FieldSpec) -> bytes:
    if field.document_type == "blob":
        payload = _require_blob_value(value, field.name)
    elif field.document_type == "int":
        integer = _require_int_value(value, field.name)
        if len(integer) > field.length:
            raise ValueError(f"{field.name} exceeds {field.length} digits")
        payload = integer.zfill(field.length).encode("ascii")
    else:
        payload = _encode_ascii(_require_string_value(value, field.name), field.name)

    prefix_length = {
        "ll_ascii": 2,
        "ll_blob": 2,
        "lll_ascii": 3,
        "lll_blob": 3,
        "llll_ascii": 4,
        "llll_blob": 4,
    }.get(field.kind, 0)
    if prefix_length:
        if len(payload) > field.length:
            raise ValueError(f"{field.name} length exceeds profile maximum")
        return f"{len(payload):0{prefix_length}d}".encode("ascii") + payload
    if len(payload) != field.length:
        raise ValueError(f"{field.name} must contain exactly {field.length} bytes")
    return payload


def _require_string_value(value: object, name: str) -> str:
    if (
        not isinstance(value, dict)
        or set(value) != {"type", "value"}
        or value.get("type") != "string"
        or not isinstance(value.get("value"), str)
    ):
        raise ValueError(f"{name} must be a closed string tagged value")
    return value["value"]


def _require_int_value(value: object, name: str) -> str:
    if (
        not isinstance(value, dict)
        or set(value) != {"type", "value"}
        or value.get("type") != "int"
        or not isinstance(value.get("value"), str)
    ):
        raise ValueError(f"{name} must be a closed int tagged value")
    integer = value["value"]
    if not integer.isascii() or not integer.isdecimal():
        raise ValueError(f"{name} must contain ASCII digits")
    if len(integer) > 1 and integer.startswith("0"):
        raise ValueError(f"{name} must use a canonical integer string")
    return integer


def _require_blob_value(value: object, name: str) -> bytes:
    if (
        not isinstance(value, dict)
        or set(value) != {"type", "value_base64"}
        or value.get("type") != "blob"
        or not isinstance(value.get("value_base64"), str)
    ):
        raise ValueError(f"{name} must be a closed blob tagged value")
    try:
        return base64.b64decode(value["value_base64"], validate=True)
    except (binascii.Error, ValueError) as error:
        raise ValueError(f"{name} must contain valid base64") from error


def _read_mti(payload: bytes) -> str:
    mti = _decode_ascii(payload, "MTI")
    _require_digits(mti, "MTI")
    return mti


def _canonical_integer(value: str, name: str) -> str:
    _require_digits(value, name)
    return value.lstrip("0") or "0"


def _require_digits(value: str, name: str) -> None:
    if not value.isascii() or not value.isdecimal():
        raise ValueError(f"{name} must contain ASCII digits")


def _decode_ascii(payload: bytes, name: str) -> str:
    try:
        return payload.decode("ascii")
    except UnicodeDecodeError as error:
        raise ValueError(f"{name} must contain ASCII bytes") from error


def _encode_ascii(value: str, name: str) -> bytes:
    try:
        return value.encode("ascii")
    except UnicodeEncodeError as error:
        raise ValueError(f"{name} must contain ASCII characters") from error


def _bitmap_has(bitmap: bytes, number: int) -> bool:
    index = number - 1
    return bool(bitmap[index // 8] & (1 << (7 - index % 8)))


def _bitmap_set(bitmap: bytearray, number: int) -> None:
    index = number - 1
    bitmap[index // 8] |= 1 << (7 - index % 8)


def _require_bytes(message: object) -> None:
    if not isinstance(message, bytes):
        raise TypeError("message must be bytes")

File: examples/external-packages/au_eftex/component/src/iso8583.rs
use base64::{Engine as _, engine::general_purpose::STANDARD};
use serde_json::{Map, Value, json};

#[derive(Clone, Copy, PartialEq)]
enum Kind {
    Fixed,
    Amount,
    Blob,
    Ll,
    Lll,
    LllBlob,
    LlllBlob,
}

#[derive(Clone, Copy)]
pub struct Field {
    pub number: usize,
    pub name: &'static str,
    kind: Kind,
    max: usize,
}

pub const FIELDS: &[Field] = &[
    Field {
        number: 0,
        name: "message_type",
        kind: Kind::Fixed,
        max: 4,
    },
    f(2, "primary_account_number", Kind::Ll, 19),
    f(3, "processing_code", Kind::Fixed, 6),
    f(4, "amount", Kind::Amount, 12),
    f(7, "transmission_time", Kind::Fixed, 10),
    f(11, "stan", Kind::Fixed, 6),
    f(12, "local_transaction_time", Kind::Fixed, 12),
    f(14, "expiration_date", Kind::Fixed, 4),
    f(22, "pos_data_code", Kind::Fixed, 15),
    f(23, "card_sequence_number", Kind::Fixed, 3),
    f(24, "function_code", Kind::Fixed, 3),
    f(25, "message_reason_code", Kind::Fixed, 4),
    f(28, "reconciliation_date", Kind::Fixed, 6),
    f(29, "reconciliation_indicator", Kind::Fixed, 3),
    f(35, "track_2_data", Kind::Ll, 37),
    f(37, "retrieval_reference_number", Kind::Fixed, 12),
    f(38, "approval_code", Kind::Fixed, 6),
    f(39, "action_code", Kind::Fixed, 3),
    f(40, "service_code", Kind::Fixed, 3),
    f(41, "terminal_id", Kind::Fixed, 8),
    f(42, "card_acceptor_id", Kind::Fixed, 15),
    f(46, "amounts_fees", Kind::Lll, 204),
    f(48, "additional_private", Kind::LlllBlob, 9999),
    f(49, "currency", Kind::Fixed, 3),
    f(50, "reconciliation_currency", Kind::Fixed, 3),
    f(52, "pin_data", Kind::Blob, 8),
    f(53, "security_control_information", Kind::Ll, 48),
    f(54, "additional_amounts", Kind::Lll, 120),
    f(55, "icc_data", Kind::LllBlob, 512),
    f(56, "original_data_elements", Kind::Ll, 31),
    f(63, "reserved_private", Kind::Lll, 800),
    f(64, "message_authentication_code", Kind::Blob, 8),
    f(74, "credits_number", Kind::Fixed, 10),
    f(75, "credits_reversal_number", Kind::Fixed, 10),
    f(76, "debits_number", Kind::Fixed, 10),
    f(77, "debits_reversal_number", Kind::Fixed, 10),
    f(81, "authorisations_number", Kind::Fixed, 10),
    f(86, "credits_amount", Kind::Fixed, 16),
    f(87, "credits_reversal_amount", Kind::Fixed, 16),
    f(88, "debits_amount", Kind::Fixed, 16),
    f(89, "debits_reversal_amount", Kind::Fixed, 16),
    f(90, "authorisations_reversal_number", Kind::Fixed, 10),
    f(97, "net_reconciliation_amount", Kind::Fixed, 17),
    f(109, "credits_fee_amounts", Kind::Ll, 84),
    f(110, "debits_fee_amounts", Kind::Ll, 84),
    f(123, "receipt_data", Kind::Lll, 999),
    f(124, "display_data", Kind::Lll, 999),
    f(128, "message_authentication_code_extended", Kind::Blob, 8),
];

const fn f(number: usize, name: &'static str, kind: Kind, max: usize) -> Field {
    Field {
        number,
        name,
        kind,
        max,
    }
}

fn field(number: usize) -> Result<&'static Field, String> {
    FIELDS
        .iter()
        .find(|field| field.number == number)
        .ok_or_else(|| format!("DE{number} is not supported by the AU EFTEX profile"))
}

fn bitmap_has(bitmap: &[u8], number: usize) -> bool {
    let index = number - 1;
    bitmap[index / 8] & (1 << (7 - index % 8)) != 0
}

fn prefix(kind: Kind) -> usize {
    match kind {
        Kind::Ll => 2,
        Kind::Lll | Kind::LllBlob => 3,
        Kind::LlllBlob => 4,
        _ => 0,
    }
}

pub fn message_length(message: &[u8]) -> Result<Option<usize>, String> {
    if message.len() < 4 {
        return Ok(None);
    }
    digits(&message[..4], "MTI")?;
    if message.len() < 12 {
        return Ok(None);
    }
    let bitmap_bytes = if bitmap_has(&message[4..12], 1) {
        16
    } else {
        8
    };
    if message.len() < 4 + bitmap_bytes {
        return Ok(None);
    }
    let bitmap = &message[4..4 + bitmap_bytes];
    if bitmap_bytes == 16 && bitmap_has(bitmap, 65) {
        return Err("tertiary bitmap is not supported by the AU EFTEX profile".into());
    }
    let mut offset = 4 + bitmap_bytes;
    for number in 2..=bitmap_bytes * 8 {
        if number == 65 || !bitmap_has(bitmap, number) {
            continue;
        }
        let spec = field(number)?;
        let prefix = prefix(spec.kind);
        if message.len() < offset + prefix {
            return Ok(None);
        }
        let length = if prefix == 0 {
            spec.max
        } else {
            digits(&message[offset..offset + prefix], spec.name)?;
            std::str::from_utf8(&message[offset..offset + prefix])
                .unwrap()
                .parse::<usize>()
                .unwrap()
        };
        if length > spec.max {
            return Err(format!("{} length exceeds profile maximum", spec.name));
        }
        offset += prefix;
        if message.len() < offset + length {
            return Ok(None);
        }
        if !matches!(spec.kind, Kind::Blob | Kind::LllBlob | Kind::LlllBlob) {
            ascii(&message[offset..offset + length], spec.name)?;
        }
        offset += length;
    }
    Ok(Some(offset))
}

pub fn decode(message: &[u8]) -> Result<Map<String, Value>, String> {
    let length = message_length(message)?.ok_or("ISO 8583 message is incomplete")?;
    if length != message.len() {
        return Err("ISO 8583 message has trailing bytes not described by bitmap".into());
    }
    let mut document = Map::new();
    document.insert(
        "message_type".into(),
        json!({"type":"string","value":ascii(&message[..4], "MTI")?}),
    );
    let bitmap_bytes = if bitmap_has(&message[4..12], 1) {
        16
    } else {
        8
    };
    let bitmap = &message[4..4 + bitmap_bytes];
    let mut offset = 4 + bitmap_bytes;
    for number in 2..=bitmap_bytes * 8 {
        if number == 65 || !bitmap_has(bitmap, number) {
            continue;
        }
        let spec = field(number)?;
        let prefix = prefix(spec.kind);
        let length = if prefix == 0 {
            spec.max
        } else {
            std::str::from_utf8(&message[offset..offset + prefix])
                .unwrap()
                .parse()
                .unwrap()
        };
        offset += prefix;
        let payload = &message[offset..offset + length];
        let value = match spec.kind {
            Kind::Blob | Kind::LllBlob | Kind::LlllBlob => {
                json!({"type":"blob","value_base64":STANDARD.encode(payload)})
            }
            Kind::Amount => {
                json!({"type":"int","value":ascii(payload,spec.name)?.trim_start_matches('0').to_string().chars().collect::<String>()})
            }
            _ => json!({"type":"string","value":ascii(payload,spec.name)?}),
        };
        let value = if spec.kind == Kind::Amount && value["value"] == "" {
            json!({"type":"int","value":"0"})
        } else {
            value
        };
        document.insert(spec.name.into(), value);
        offset += length;
    }
    Ok(document)
}

pub fn encode(document: &Map<String, Value>) -> Result<Vec<u8>, String> {
    for name in document.keys() {
        if !FIELDS.iter().any(|field| field.name == name) {
            return Err(format!("unknown document field: {name}"));
        }
    }
    let mti = tagged_text(
        document
            .get("message_type")
            .ok_or("message_type is required")?,
        "message_type",
        "string",
    )?;
    if mti.len() != 4 {
        return Err("message_type must contain exactly 4 ASCII digits".into());
    }
    digits(mti.as_bytes(), "message_type")?;
    let present: Vec<_> = FIELDS[1..]
        .iter()
        .filter(|field| document.contains_key(field.name))
        .collect();
    let secondary = present.iter().any(|field| field.number > 64);
    let mut bitmap = vec![0u8; if secondary { 16 } else { 8 }];
    if secondary {
        bitmap[0] |= 0x80;
    }
    let mut output = mti.into_bytes();
    for spec in &present {
        let index = spec.number - 1;
        bitmap[index / 8] |= 1 << (7 - index % 8);
    }
    output.extend(bitmap);
    for spec in present {
        let value = &document[spec.name];
        let mut payload = match spec.kind {
            Kind::Blob | Kind::LllBlob | Kind::LlllBlob => tagged_blob(value, spec.name)?,
            Kind::Amount => {
                let integer = tagged_text(value, spec.name, "int")?;
                if !integer.bytes().all(|byte| byte.is_ascii_digit())
                    || (integer.len() > 1 && integer.starts_with('0'))
                {
                    return Err(format!("{} must use a canonical integer string", spec.name));
                }
                format!("{integer:0>width$}", width = spec.max).into_bytes()
            }
            _ => tagged_text(value, spec.name, "string")?.into_bytes(),
        };
        if payload.len() > spec.max {
            return Err(format!("{} length exceeds profile maximum", spec.name));
        }
        let prefix = prefix(spec.kind);
        if prefix == 0 && payload.len() != spec.max {
            return Err(format!(
                "{} must contain exactly {} bytes",
                spec.name, spec.max
            ));
        }
        if prefix != 0 {
            output.extend(format!("{:0width$}", payload.len(), width = prefix).bytes());
        }
        output.append(&mut payload);
    }
    Ok(output)
}

fn tagged_text(value: &Value, name: &str, kind: &str) -> Result<String, String> {
    let object = value
        .as_object()
        .ok_or_else(|| format!("{name} must be a closed {kind} tagged value"))?;
    if object.len() != 2 || object.get("type").and_then(Value::as_str) != Some(kind) {
        return Err(format!("{name} must be a closed {kind} tagged value"));
    }
    let text = object
        .get("value")
        .and_then(Value::as_str)
        .ok_or_else(|| format!("{name} must be a closed {kind} tagged value"))?;
    if !text.is_ascii() {
        return Err(format!("{name} must contain ASCII characters"));
    }
    Ok(text.to_owned())
}

fn tagged_blob(value: &Value, name: &str) -> Result<Vec<u8>, String> {
    let object = value
        .as_object()
        .ok_or_else(|| format!("{name} must be a closed blob tagged value"))?;
    if object.len() != 2 || object.get("type").and_then(Value::as_str) != Some("blob") {
        return Err(format!("{name} must be a closed blob tagged value"));
    }
    STANDARD
        .decode(
            object
                .get("value_base64")
                .and_then(Value::as_str)
                .ok_or_else(|| format!("{name} must be a closed blob tagged value"))?,
        )
        .map_err(|_| format!("{name} must contain valid base64"))
}

fn digits(value: &[u8], name: &str) -> Result<(), String> {
    if value.iter().all(u8::is_ascii_digit) {
        Ok(())
    } else {
        Err(format!("{name} must contain ASCII digits"))
    }
}
fn ascii<'a>(value: &'a [u8], name: &str) -> Result<&'a str, String> {
    std::str::from_utf8(value)
        .map_err(|_| format!("{name} must contain ASCII bytes"))
        .and_then(|value| {
            if value.is_ascii() {
                Ok(value)
            } else {
                Err(format!("{name} must contain ASCII bytes"))
            }
        })
}

pub fn contains_mac(document: &Map<String, Value>) -> bool {
    document.contains_key("message_authentication_code")
        || document.contains_key("message_authentication_code_extended")
}
